VariantShredding.md

```diff
@@ -25,276 +25,302 @@
+The series of `tags` arrays `["comedy", "drama"], ["horror", null], ["comedy", "drama", "romance"], null` would be stored as:
 
-# Shredding Semantics
+| Array                            | `value`     | `typed_value `| `typed_value...value` | `typed_value...typed_value`    |
+|----------------------------------|-------------|---------------|-----------------------|--------------------------------|
+| `["comedy", "drama"]`            | null        | non-null      | [null, null]          | [`comedy`, `drama`]            |
+| `["horror", null]`               | null        | non-null      | [null, `00`]          | [`horror`, null]               |
+| `["comedy", "drama", "romance"]` | null        | non-null      | [null, null, null]    | [`comedy`, `drama`, `romance`] |
+| null                             | `00` (null) | null          |                       |                                |
 
-Reconstruction of Variant value from a shredded representation is not expected to produce a bit-for-bit identical binary to the original unshredded value.
-For example, the order of fields in the binary may change, as may the physical representation of scalar values.
+#### Objects
 
-In particular, the [Variant Binary Encoding](VariantEncoding.md) considers all integer and decimal representations to represent a single logical type.
-As a result, it is valid to shred a decimal into a decimal column with a different scale, or to shred an integer as a decimal, as long as no numeric precision is lost.
-For example, it would be valid to write the value 123 to a Decimal(9, 2) column, but the value 1.234 would need to be written to the **variant_value** column.
-When reconstructing, it would be valid for a reader to reconstruct 123 as an integer, or as a Decimal(9, 2).
-Engines should not depend on the physical type of a Variant value, only the logical type.
+Fields of an object can be shredded using a Parquet group for `typed_value` that contains shredded fields.
 
-On the other hand, shredding as a different logical type is not allowed.
-For example, the integer value 123 could not be shredded to a string `typed_value` column as the string "123", since that would lose type information.
-It would need to be written to the `variant_value` column.
+If the value is not an object, `typed_value` must be null.
 
-# Reconstructing a Variant
+If the value is a partially shredded object, the `value` must not contain the shredded fields.
+If shredded fields are present in the variant object, it is invalid and readers must either fail or use the shredded values.
 
-It is possible to recover a full Variant value using a recursive algorithm, where the initial call is to `ConstructVariant` with the top-level fields, which are assumed to be null if they are not present in the schema.
+Each shredded field in the `typed_value` group is represented as a required group that contains optional `value` and `typed_value` fields.
```

Comment:
I think that here and in the array case, it would be good to clarify whether `typed_value` can be omitted from the schema entirely. E.g. if there's no consistent type for a field, I think we'd still want to shred the field, but put all values in `value`, and not require that a `typed_value` type be specified.

Reply:
Conversely, is `value` always required? Would it be valid for a writer to only create a `typed_value` column if it knows that all values have a predictable type that can be shredded?

Reply:
I think that `typed_value` could be omitted. Wouldn't that be the case where the element is simply not shredded? I think we should call it out here, but I'll also adjust the language so that it is clear that shredding is not required, except for object fields, where you'd simply not have a shredded field (it makes no sense to shred a field and not include `typed_value`).

Reply:
Agreed, I think it makes complete sense for `typed_value` to be omitted, but it would be good to be clear.

I'm less clear about whether it should be valid (here, or for other types) to omit `value`, and treat that as equivalent to all null. I can sort of imagine cases where you're converting a typed parquet schema to variant, and know the types well enough to know that `value` will never be needed, but it seems like a fairly marginal benefit to omit `value` from the schema vs. leaving it and populating it will all nulls. Assuming that we don't want to allow that, it might be good to clarify somewhere that the `value` column is always required.

Reply:
I do not get the benefit of optionally omitting typed_value. I think Ryan said the same thing above but omitting typed_value should be equivalent to not shredding a field in which case field should not exist in the schema to begin with. Omitting typed_value just seems to increase the number of cases to deal with. Similar argument can be made for value but actually being able to omit value has a benefit. If a writer uses V1 Parquet pages where you cannot get number of null values before reading def lvls for a field, not having value field in the schema for a perfectly shredded field would allow readers skip trying to figure out values are all null after reading the def lvls.

Reply:
If a field has an inconsistent type, it may still be useful to shred it into `value` in order to fetch without the extra IO required to get the rest of the Variant.

> not having value field in the schema for a perfectly shredded field would allow readers skip trying to figure out values are all null after reading the def lvls

Couldn't it figure this out from the row group stats, since `value` should be null for all rows if it could have been omitted?

Reply:
Yes, I agree shredding into value in case of inconsistent types is useful and I also really like that the changes to the spec makes is really clear as to when/how that happens.

Regarding value field, yes we could figure that out from row group stats but that is not always present.

Reply:
Added this:
> The `typed_value` field may be omitted when not shredding elements as a specific type.
> When `typed_value` is omitted, `value` must be `required`.

I think there is value in allowing elements to be shredded. We could get dictionary encoding for them.